op-node: Skip L2 Genesis Check when EL syncing

trianglesphere opened the pull request:
**Description**

This is done because the transition block does not exist in a geth when
performing a snap sync and the actual genesis block header is pre-london
which breaks all of our fetching code.

This change is required for snap sync to work without downloading a datadir on transitioned networks.

**Tests**

Tested that this works with an empty datadir on goerli.


**Metadata***

* Fixes https://github.com/ethereum-optimism/client-pod/issues/376

protolambda commented on op-service/sources/types.go:
```diff
@@ -125,6 +125,11 @@ type rpcHeader struct {
 // checkPostMerge checks that the block header meets all criteria to be a valid ExecutionPayloadHeader,
 // see EIP-3675 (block header changes) and EIP-4399 (mixHash usage for prev-randao)
 func (hdr *rpcHeader) checkPostMerge() error {
+	// Skip checking if a block is post-merge for genesis blocks.
+	// We need this when validating the genesis block for op-goerli & op-mainnet in CheckL2GenesisBlockHash
+	if hdr.Number == 0 {
+		return nil
+	}
```
This also disables it for real post-merge networks that have a valid post-merge genesis. Instead, I recommend setting `MustBePostMerge: config.Genesis.L2.Number == 0,` in the config, so we can accept pre-merge headers, but only when necessary.

trianglesphere replied:
But `MustBePostMerge: config.Genesis.L2.Number == 0,` disables the check for all blocks even when the only pre-merge block we get is this genesis. Also I don't think we every pull any pre-merge L1 block

protolambda replied:
Maybe check as `hdr.Number == 0 && config.Genesis.L2.Number != 0` then?

ajsutton replied:
There's no reason that a merged-at-genesis chain has to have a difficulty of 0 in its genesis header.  Typically it does and I can see how EIP-3675 says it should be but I'm not aware of any EL that would validate it.  The genesis block just is and can be wildly invalid in a wide range of ways - the only thing that matters is that everyone has the same genesis block.

So I'd actually be ok with skipping checks for all genesis blocks given that we're checking it matches the block hash anyway.

protolambda commented on op-node/rollup/types.go:
```diff
@@ -205,11 +205,21 @@ func (cfg *Config) CheckL2ChainID(ctx context.Context, client L2Client) error {
 // CheckL2GenesisBlockHash checks that the configured L2 genesis block hash is valid for the given client.
 func (cfg *Config) CheckL2GenesisBlockHash(ctx context.Context, client L2Client) error {
-	l2GenesisBlockRef, err := client.L2BlockRefByNumber(ctx, cfg.Genesis.L2.Number)
+	l2Number := cfg.Genesis.L2.Number
+	l2Hash := cfg.Genesis.L2.Hash
+	// Use the actual genesis on transitioned networks (i.e. op-mainnet & op-goerli).
+	if cfg.L2ChainID.Uint64() == 10 {
+		l2Number = 0
+		l2Hash = common.HexToHash("0x7ca38a1916c42007829c55e69d3e9a73265554b586a499015373241b8a3fa48b")
+	} else if cfg.L2ChainID.Uint64() == 420 {
+		l2Number = 0
+		l2Hash = common.HexToHash("0xc1fc15cd51159b1f1e5cbc4b82e85c1447ddfa33c52cf1d98d14fba0d6354be1")
+	}
```
There are chainID constants in this package you could use instead: `opMainnet` and `opGoerli`

protolambda commented on op-node/rollup/types.go:
```diff
@@ -205,11 +205,21 @@ func (cfg *Config) CheckL2ChainID(ctx context.Context, client L2Client) error {
 	}
-	if l2GenesisBlockRef.Hash != cfg.Genesis.L2.Hash {
+	if l2GenesisBlockRef.Hash != l2Hash {
 		return fmt.Errorf("incorrect L2 genesis block hash %s, expected %s", l2GenesisBlockRef.Hash, cfg.Genesis.L2.Hash)
```
good rabbit

tynes commented:
@trianglesphere Is this required for ecotone to work on OP Mainnet?

trianglesphere commented:
@tynes yes b/c this is required for snap sync. It's on my list to address the PR comments here.

trianglesphere committed:
op-node: Skip L2 Genesis Check when EL syncing

This is done because the transition block does not exist in a geth when
performing a snap sync and the actual genesis block header is pre-london
which breaks all of our fetching code.

trianglesphere commented on op-node/rollup/types.go:
```diff
@@ -128,13 +128,17 @@ func (cfg *Config) ValidateL1Config(ctx context.Context, client L1Client) error
 	}
 
-	// Validate the Rollup L2 Genesis Blockhash
+	// Validate the Rollup L2 Genesis Blockhash if requested
+	// We skip this when doing EL sync
```
```suggestion
	// Validate the Rollup L2 Genesis Blockhash if requested. We skip this when doing EL sync
```

trianglesphere committed:
Update op-node/rollup/types.go

protolambda reviewed (approved):
I think it would be nice to check it for chains with an op-stack bedrock genesis, but disabling it entirely in case of EL sync is ok too.
